### ISSUE TITLE
Validation for function section and code section in binary spec

### ISSUE THREAD

**Issue by rianhunter** [WebAssembly/spec#990]
To find out the validation rules for WebAssembly, I usually refer to the validation section of the spec. The validation rule that the number of entries in the function section and the code section are the same are in a different location:

https://github.com/WebAssembly/spec/blame/master/document/core/binary/modules.rst#L424

This is slightly not intuitive. When I wrote my parser I didn't include that rule, and when I wrote my validator I didn't include that rule. This is just a humble suggestion but maybe specify this more boldly in the binary section of the spec, and repeat it as a note in the validation section (Even though it only deals with abstract WebAssembly structures not the binary format).

**Comment by binji** [comment]
I see what you mean, though it seems like it would be odd to talk about the `function` and `code` sections in the validation chapter. It's also worth noting that these cases are not considered "invalid", but "malformed" by the testsuite. See the tests here: https://github.com/WebAssembly/spec/blob/master/test/core/binary.wast#L607

**Comment by rianhunter** [comment]
Yeah I know they are malformed, that's why my parser has the check. Admittedly an error on my part for not catching that line when I first wrote the parser. Feel free to close if you don't think this is actionable feedback.

**Comment by binji** [comment]
Sorry, I just meant that the spec and the tests differentiate between this being a validation error and a binary decoding error. In any case, I do think that the constraint is subtle and could be highlighted more. @rossberg WDYT?

**Comment by rossberg** [comment]
@binji, do you have a suggestion? The binary format spec already says:

> The lengths of vectors produced by the (possibly empty) :ref:`function <binary-funcsec>` and :ref:`code <binary-codesec>` section must match up.

**Comment by binji** [comment]
I just thought it might be worth calling out in the spec more, maybe just adding a paragraph break.

**Comment by rossberg** [comment]
Good idea, paragraph break added in #1001.
